File: backend/app/api/routes/stones.py
from fastapi import APIRouter, Query

from app.api.deps import DbSession, UserId
from app.schemas.common import ApiResponse
from app.schemas.stone import (
    CreateStoneRequest,
    CreateStoneResponse,
    StoneListResponse,
    StoneResponse,
)
from app.services.stone_service import StoneService

router = APIRouter(prefix="/stones", tags=["stones"])


@router.post("", response_model=ApiResponse[CreateStoneResponse])
async def create_stone(request: CreateStoneRequest, db: DbSession, user_id: UserId):
    service = StoneService(db)
    result = await service.create_stone(user_id, request)
    return ApiResponse(data=result)


@router.get("", response_model=ApiResponse[StoneListResponse])
async def list_stones(
    db: DbSession,
    user_id: UserId,
    page: int = Query(1, ge=1),
    page_size: int = Query(50, ge=1, le=100),
    type: str | None = Query(None),
):
    service = StoneService(db)
    result = await service.list_stones(user_id, page, page_size, type)
    return ApiResponse(data=result)


@router.get("/{stone_id}", response_model=ApiResponse[StoneResponse])
async def get_stone(stone_id: str, db: DbSession, user_id: UserId):
    service = StoneService(db)
    result = await service.get_stone(user_id, stone_id)
    return ApiResponse(data=result)


@router.post("/{stone_id}/favorite")
async def favorite_stone(stone_id: str, db: DbSession, user_id: UserId):
    service = StoneService(db)
    await service.toggle_favorite(user_id, stone_id)
    return ApiResponse(data={"message": "ok"})


@router.delete("/{stone_id}/favorite")
async def unfavorite_stone(stone_id: str, db: DbSession, user_id: UserId):
    service = StoneService(db)
    await service.toggle_favorite(user_id, stone_id)
    return ApiResponse(data={"message": "ok"})


@router.delete("/{stone_id}")
async def delete_stone(stone_id: str, db: DbSession, user_id: UserId):
    service = StoneService(db)
    await service.delete_stone(user_id, stone_id)
    return ApiResponse(data={"message": "deleted"})

File: frontend/lib/api/stones.ts
import { apiRequest, type ApiResponse } from "@/lib/api/client";
import type { CreateStoneRequest, Stone, StoneListResponse } from "@/types/stone";

interface CreateStoneData {
  stone: Stone;
  hp_change: number;
  hp_after: number;
}

export async function createStone(req: CreateStoneRequest): Promise<ApiResponse<CreateStoneData>> {
  return apiRequest<CreateStoneData>("/stones", {
    method: "POST",
    body: JSON.stringify(req),
  });
}

export async function listStones(
  page = 1,
  pageSize = 50,
  type?: string
): Promise<ApiResponse<StoneListResponse>> {
  const params = new URLSearchParams({
    page: String(page),
    page_size: String(pageSize),
  });
  if (type) params.set("type", type);
  return apiRequest<StoneListResponse>(`/stones?${params}`);
}

export async function getStone(id: string): Promise<ApiResponse<Stone>> {
  return apiRequest<Stone>(`/stones/${id}`);
}

export async function toggleFavorite(id: string, on: boolean): Promise<ApiResponse<Stone>> {
  return apiRequest<Stone>(`/stones/${id}/favorite`, {
    method: on ? "POST" : "DELETE",
  });
}

export async function deleteStone(id: string): Promise<ApiResponse<null>> {
  return apiRequest<null>(`/stones/${id}`, { method: "DELETE" });
}
